graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java

```diff
@@ -0,0 +1,158 @@
+ * Lock expiry is handled by MongoDB internally. We set lock expiry to 60 seconds, but in practice it may take up to ~2
+ * minutes until a lock really expires.
+ */
+@Singleton
+public class MongoLockService implements LockService {
+
+    public static final String COLLECTION_NAME = "cluster_locks";
+    public static final java.time.Duration MIN_LOCK_TTL = Duration.ofSeconds(60);
+
+    private final NodeId nodeId;
+    private final MongoCollection<Document> collection;
+
+    @Inject
+    public MongoLockService(NodeId nodeId,
+                            MongoConnection mongoConnection,
+                            @Named("leader_election_lock_ttl") Duration leaderElectionLockTTL) {
+        this.nodeId = nodeId;
+
+        collection = mongoConnection.getMongoDatabase().getCollection(COLLECTION_NAME);
+
+        collection.createIndex(Indexes.ascending(FIELD_RESOURCE_NAME), new IndexOptions().unique(true));
+
+        final Bson updatedAtKey = Indexes.ascending(FIELD_UPDATED_AT);
+        final IndexOptions indexOptions = new IndexOptions().expireAfter(leaderElectionLockTTL.getSeconds(), TimeUnit.SECONDS);
+        ensureTTLIndex(collection, updatedAtKey, indexOptions);
+    }
+
+    // MongoDB Indexes cannot be altered once created. If the leaderElectionLockTTL has changed, replace the index
+    private void ensureTTLIndex(MongoCollection<Document> collection, Bson updatedAtKey, IndexOptions indexOptions) {
+        for (Document document : collection.listIndexes()) {
+            final Set<String> keySet = document.get("key", Document.class).keySet();
+            if (keySet.contains(FIELD_UPDATED_AT)) {
+                // Since MongoDB 5.0 this is an Integer. Used to be a Long ¯\_(ツ)_/¯
+                final long expireAfterSeconds = document.get("expireAfterSeconds", Number.class).longValue();
+                if (Objects.equals(expireAfterSeconds, indexOptions.getExpireAfter(TimeUnit.SECONDS))) {
+                    return;
+                }
+                collection.dropIndex(updatedAtKey);
+            }
+        }
+        // not found or dropped, creating new index
+        collection.createIndex(updatedAtKey, indexOptions);
+    }
+
+    /**
+     * Request a lock. If a lock already exists, the lock expiry time will be
+     *
+     * @param resourceName Unique identifier for the resource that should be guarded by this lock.
+     * @return A {@link Lock} object, if a lock was obtained. An empty {@link Optional}, if no lock could be acquired.
+     */
+    @Override
+    public Optional<Lock> lock(@Nonnull String resourceName) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(resourceName));
+
+        try {
+            final Document doc = collection.findOneAndUpdate(
+                    and(eq(FIELD_RESOURCE_NAME, resourceName), eq(FIELD_LOCKED_BY, nodeId.toString())),
+                    currentDate(FIELD_UPDATED_AT),
```

Comment:
This confused me a lot and has cost me some time during the review. :sweat_smile: 

My first thought was: we don't set `resource` and `locked_by` in the update operation. How are these fields event set?

After lots of digging in the source code and MongoDB documentation I found it in the docs:

> If no document matches the query criteria and the `<update>` parameter is a document with update operator expressions, then the operation creates a base document from the **equality clauses** in the `<query>` parameter and applies the expressions from the `<update>` parameter.

https://docs.mongodb.com/manual/reference/method/db.collection.update/#upsert-with-operator-expressions---set-

Not sure if that's common knowledge, but it was highly confusing for me. What do you think about adding a comment based on the documentation snippet above?

Alternatively we could explicitly set the `resource` and `locket_by` in the update operation to make it clearer.

What do you think?

Reply:
Sorry to cause the confusion. In that case, we should make this more obvious. Comment or `setOnInsert` are both viable options.